File: Blind75/imp interview/valid parentheses.py
'''
Concept:
using stack to store and pop characters
hasp map stores - closed brackets as keys with open brackets as values
when a closed bracket is encountered check haspmap for value and match them with going through stack


memory & time = O(n)
'''

class Solution:
    def isValid(self, s: str) -> bool:
        stackList = []
        bracketHM = { ")" : "(", "}" : "{",  "]" : "[" }

        for i in s:
            if i in bracketHM:
                if stackList and stackList[-1] == bracketHM[i]:
                    stackList.pop()
                else:
                    return False
            else:
                stackList.append(i) # if you open bracket is encountered
        
        return True if not stackList else False # only enter true is stack is empty

File: grind75/Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
    unordered_map<char, char> map = { {')', '('}, {']', '['}, {'}', '{'} };
    stack<char> stack;
    for (int c : s) {
        if (map.count(c)) { // check if c is closing bracket - returns true if close exists as a key
            if (stack.top() == map[c]) { // check if top of stack is open bracket
                stack.pop();
            }
            else {
                return false;
            }
        }
        else { // then its a opening bracket so push to stack
            stack.push(c);
        }
    }
    return stack.empty(); // returns true only is stack has no elements
    }
};